File: backend/app/routers/categories.py
from fastapi import APIRouter, Depends, HTTPException, Response
from sqlalchemy.orm import Session

from app.database.dependencies import get_db
from app.models.category import Category
from app.schemas.category import CategoryCreate, CategoryUpdate, CategoryRead

router = APIRouter(
    prefix="/categories",
    tags=["Categories"]
)

@router.post("/", response_model=CategoryRead, status_code=201)
def create_category(
    category_data: CategoryCreate,
    db: Session = Depends(get_db),
):
    category = Category(
        name=category_data.name,
        description=category_data.description,
        target_name=category_data.target_name,
        target_type=category_data.target_type,
    )

    db.add(category)
    db.commit()
    db.refresh(category)

    return category

@router.get("/", response_model=list[CategoryRead])
def get_categories(
    db: Session = Depends(get_db),
):
    categories = db.query(Category).all()
    return categories

@router.get("/{category_id}", response_model=CategoryRead)
def get_category(
    category_id: int,
    db: Session = Depends(get_db),
):
    category = (
        db.query(Category)
        .filter(Category.id == category_id)
        .first()
    )

    if category is None:
        raise HTTPException(
            status_code=404,
            detail="Category not found",
        )

    return category

@router.put("/{category_id}", response_model=CategoryRead)
def update_category(
    category_id: int,
    category_data: CategoryUpdate,
    db: Session = Depends(get_db),
):
    category = (
        db.query(Category)
        .filter(Category.id == category_id)
        .first()
    )

    if category is None:
        raise HTTPException(
            status_code=404,
            detail="Category not found",
        )

    update_data = category_data.model_dump(exclude_unset=True)

    for field, value in update_data.items():
        setattr(category, field, value)

    db.commit()
    db.refresh(category)

    return category

@router.delete("/{category_id}", status_code=204)
def delete_category(
    category_id: int,
    db: Session = Depends(get_db),
):
    category = (
        db.query(Category)
        .filter(Category.id == category_id)
        .first()
    )

    if category is None:
        raise HTTPException(
            status_code=404,
            detail="Category not found",
        )

    db.delete(category)
    db.commit()

    return Response(status_code=204)

File: frontend/src/api/Categories.ts
import request from "./Client";

export interface Category {
  id: number;

  name: string;

  description: string;

  target_name: string;

  target_type: string;
}

export interface CategoryCreate {
  name: string;

  description: string;

  target_name: string;

  target_type: string;
}

export type CategoryUpdate = Partial<CategoryCreate>;

export async function getCategories(): Promise<Category[]> {
  return request<Category[]>("/categories/");
}

export async function getCategory(id: number): Promise<Category> {
  return request<Category>(`/categories/${id}`);
}

export async function createCategory(data: CategoryCreate): Promise<Category> {
  return request<Category>("/categories/", {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export async function updateCategory(
  id: number,
  data: CategoryUpdate,
): Promise<Category> {
  return request<Category>(`/categories/${id}`, {
    method: "PUT",
    body: JSON.stringify(data),
  });
}

export async function deleteCategory(id: number): Promise<void> {
  await request<void>(`/categories/${id}`, {
    method: "DELETE",
  });
}
